modules/imap/functions.php

```diff
@@ -1673,3 +1673,722 @@ function is_imap_archive_folder($server_id, $user_config, $current_folder) {
+}}
+
+/**
+ * Report spam message to SpamCop
+ * Uses authenticated SMTP to ensure proper SPF/DKIM validation
+ * Must use the exact email address from the IMAP server where the message is located
+ */
+if (!hm_exists('report_spam_to_spamcop')) {
+function report_spam_to_spamcop($message_source, $reasons, $user_config, $session = null, $imap_server_email = '') {
+    $spamcop_enabled = $user_config->get('spamcop_enabled_setting', false);
+    if (!$spamcop_enabled) {
+        return array('success' => false, 'error' => 'SpamCop reporting is not enabled');
+    }
+
+    $spamcop_email = $user_config->get('spamcop_submission_email_setting', '');
+    if (empty($spamcop_email)) {
+        return array('success' => false, 'error' => 'SpamCop submission email not configured');
+    }
+
+    $sanitized_message = sanitize_message_for_spam_report($message_source, $user_config);
+
+    // SpamCop requires the exact email address associated with the account
+    $from_email = '';
+    if (!empty($imap_server_email)) {
+        $from_email = $imap_server_email;
+    } else {
+        $from_email = $user_config->get('spamcop_from_email_setting', '');
+        if (empty($from_email)) {
+            $imap_servers = $user_config->get('imap_servers', array());
+            if (!empty($imap_servers)) {
+                $first_server = reset($imap_servers);
+                $from_email = isset($first_server['user']) ? $first_server['user'] : '';
+            }
+        }
+    }
+
+    if (empty($from_email)) {
+        return array('success' => false, 'error' => 'No sender email address configured');
+    }
+
+    $subject = 'Spam report';
+
+    if (!class_exists('Hm_MIME_Msg')) {
+        $mime_file = (defined('APP_PATH') ? APP_PATH : dirname(__FILE__) . '/../') . 'modules/smtp/hm-mime-message.php';
+        if (file_exists($mime_file)) {
+            require_once $mime_file;
+        } else {
+            return array('success' => false, 'error' => 'SMTP module required for SpamCop reporting. Please enable the SMTP module.');
+        }
+    }
+    
+    // temporary file for the spam message attachment
+    $temp_file = create_spam_report_temp_file($sanitized_message, $user_config, $session, 'spamcop_');
+    
+    $body = '';
+    $mime = new Hm_MIME_Msg($spamcop_email, $subject, $body, $from_email, false, '', '', '', '', $from_email);
+    
+    $attachment = array(
+        'name' => 'spam.eml',
+        'type' => 'message/rfc822',
+        'size' => strlen($sanitized_message),
+        'filename' => $temp_file
+    );
+    
+    $mime->add_attachments(array($attachment));
+
+    $mime_message = $mime->get_mime_msg();
+    
+    // SpamCop rejects automated submissions, so removed X-Mailer headers
+    $mime_message = preg_replace('/^X-Mailer:.*$/mi', '', $mime_message);
+    $mime_message = preg_replace('/\r\n\r\n+/', "\r\n\r\n", $mime_message); // Clean up extra blank lines
+    
+    $encoding_result = fix_spam_report_encoding($mime_message, 'SpamCop');
```

Comment:
You could probably use the right encoding in get_mime_msg above and spare the "fixing" part here which seems hacky...